#170947548 Login with Facebook and Google

Ntare22 opened the pull request:
#### What does this PR do?
- Enables users to sign with their google or facebook accounts

#### Description of Task to be completed?
- Install passport, passport-facebook, and passport-google-oauth
- Setup apps on google developer console and facebook facebook developer
- Write passport configuration file with facebook and google
- create a getProfile callback function.
- Setup routes for google and facebook redirecting
- Initialise passport in index.js

#### How should this be manually tested?
- Clone repository 
- Run `npm install`
- Open localhost link provided in terminal. ex `http://localhost:8000/api`
- Then open this link `http://localhost:8000/api/v1/auth/google` or `facebook`
#### Any background context you want to provide?
- N/A
#### What are the relevant pivotal tracker stories?
[#170947548](https://www.pivotaltracker.com/story/show/170947548)
#### Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/30793016/74716440-3e4af480-5237-11ea-997a-da554fece49b.png)
![image](https://user-images.githubusercontent.com/30793016/74725899-3a26d300-5247-11ea-86c7-8414d17cc186.png)
![image](https://user-images.githubusercontent.com/30793016/74725910-3f841d80-5247-11ea-97d0-fa87e89a992e.png)

Ntare22 committed:
ft(Login with Facebook and Google):

Install passport, passport-facebook, and passport-google-oauth
Setup apps on google developer console and facebook facebook developer
Write passport configuration file with facebook and google
create a getProfile callback function.
Setup routes for google and facebook redirecting
Initialise passport in index.js

[ #170947548 in progress]

Mnickii commented on src/routes/authRoutes.js:
```diff
@@ -1,13 +1,32 @@
+
+
+authRouter.get('/facebook', passport.authenticate('facebook', {
+  scope: ['emails'],
```
why do we have this scope?

Ntare22 committed:
minor changes to the facebook oauth

Ntare22 committed:
fixed facebook

Ntare22 committed:
fixed tests

Ntare22 committed:
fixed some bugs in the controller
wrote tests for the controller

[ Finishes #170947548 ]

 Please enter the commit message for your changes. Lines starting

Mnickii commented on src/seeders/20200212121323-User.js:
```diff
@@ -4,8 +4,17 @@ module.exports = {
+      //   firstName: 'bienaime',
+      //   lastName: 'jeanb',
+      //   email: 'aime@andela.com',
+      //   password: bcrypt.hashSync('Aime12&*', Number(process.env.passwordHashSalt)),
```
remove commented code

Mnickii commented on src/controllers/authController.js:
```diff
@@ -89,6 +89,53 @@ export default class AuthController {
+      const isVerified = true;
+      const password = 'null';
+
+      // const email = Array.isArray(emails) && emails[0] && emails[0].value;
```
commented code

Mnickii commented on src/controllers/authController.js:
```diff
@@ -89,6 +89,52 @@ export default class AuthController {
+      const { id } = profile;
+      const firstName = profile.name.givenName;
+      const lastName = profile.name.familyName;
+      const email = Array.isArray(profile.emails) ? profile.emails[0].value : '';
```
Do we need to check specifically if it's an array or if the profile.emails are present? If the latter, ```profile.emails ? profile.emails[0].value : ''``` should do

Ntare22 replied:
I changed that.

Mnickii commented on src/seeders/20200212121323-User.js:
```diff
@@ -6,7 +6,7 @@ module.exports = {
     [
       {
-        id: uuid(),
+        id: "d0551d02-46e6-4b79-9c3b-8d378c178243",
```
why this change?

Ntare22 replied:
It Baraka who put the uuid().... which might mean I didn't rebase well.

Mnickii replied:
Oh, that should be it. It seemed as though you'd undone Baraka's changes

Mnickii commented on src/tests/oAuth.test.js:
```diff
@@ -0,0 +1,86 @@
+  it('should return user is successfully created', (done) => {
+    chai
+      .request(app)
+      .post('/api/v1/auth/facebook')
```
it's not efficient to send unwanted requests to external APIs. Remember we have rate limiting and the way you've done it, we hit the external APIs everytime someone runs the tests

Mnickii commented on src/tests/oAuth.test.js:
```diff
@@ -0,0 +1,106 @@
+    const results = await authController.oAuthLogin(req, res);
+    expect(results.body.status).to.equal(200);
+    expect(results.body.message).to.equal('User is successfully created');
+    // console.log('>>>', results);
```
commented code.

Mnickii commented on src/tests/index.test.js:
```diff
@@ -20,16 +19,3 @@ describe('Testing index page ', () => {
 });
-
-describe('Testing index page ', () => {
-  it('should return a message from index page in French.', (done) => {
```
why remove this test?

Ntare22 replied:
It is not removed.... we rename the testing documents...!

Mnickii commented on src/tests/4-verifyAccount.test.js:
```diff
@@ -43,7 +43,7 @@ describe('Verification tests', () => {
       .get(`/api/v1/auth/verification?token=${wrongToken}&email=${email}`)
       .end((err, res) => {
-        expect(res.body).to.be.an('object');
+        expect(res.body).to.be.a('object');
```
this change isn't necessary

Mnickii commented on src/tests/5-oAuth.test.js:
```diff
@@ -0,0 +1,106 @@
+  it('should redirect to facebook', async () => {
+    chai
+      .request(index)
+      .get('/api/v1/auth/facebook')
```
let's not hit external APIs at all. mock the requests

Mnickii commented on src/tests/1-index.test.js:
```diff
@@ -20,16 +19,3 @@ describe('Testing index page ', () => {
-      .get('/api?lang=fr')
-      .end((err, res) => {
-        expect(res.body).to.be.an('object');
-        expect(res.status).to.equal(200);
```
this test is necessary

Mnickii commented on src/tests/5-oAuth.test.js:
```diff
@@ -0,0 +1,106 @@
+    const results = await authController.oAuthLogin(req, res);
+    expect(results.body.status).to.equal(200);
+    expect(results.body.message).to.equal('User is successfully created');
+    // console.log('>>>', results);
```
remove commented code

Mnickii commented on src/services/factory/googleUser.js:
```diff
@@ -0,0 +1,10 @@
+import Chance from 'chance';
+import { Factory } from 'rosie';
+
+const chance = new Chance();
```
we don't need these anymore

Mnickii commented on src/services/factory/facebookUser.js:
```diff
@@ -0,0 +1,10 @@
+const chance = new Chance();
+
+export default Factory.define('user')
+  .attr('id', `${chance.integer({ min: 100, max: 100000 })}`)
```
this too

Mnickii commented on package.json:
```diff
@@ -33,17 +34,23 @@
     "i18n": "^0.8.5",
     "jsonwebtoken": "^8.5.1",
     "localStorage": "^1.0.4",
+    "node-mocks-http": "^1.8.1",
```
is this used somewhere? let's not include unused packages,

Ntare22 replied:
I've removed it.

Mnickii commented on src/tests/3-user.test.js:
```diff
@@ -21,7 +21,7 @@ describe('USER ROLES TESTS', () => {
       .end((err, res) => {
         token = res.body.data;
-        expect(res.status).to.equal(200);
+        expect(res.status).equal(200);
```
return this

Mnickii commented on src/tests/3-user.test.js:
```diff
@@ -36,7 +36,7 @@ describe('USER ROLES TESTS', () => {
         expect(res.status).to.equal(200);
         unauthToken = res.body.data;
-
+        expect(res.status).equal(200);
```
remove this

Mnickii reviewed (approved):
LGTM
